File: graphs/binary search concept/Tree Path.go
package main

import (
	"bufio"
	"fmt"
	"os"
)

type FastScanner struct {
	r *bufio.Reader
}

func NewFastScanner() *FastScanner {
	return &FastScanner{
		r: bufio.NewReaderSize(os.Stdin, 1<<20),
	}
}

func (fs *FastScanner) NextInt() int {
	sign, val := 1, 0

	c, _ := fs.r.ReadByte()
	for (c < '0' || c > '9') && c != '-' {
		c, _ = fs.r.ReadByte()
	}

	if c == '-' {
		sign = -1
		c, _ = fs.r.ReadByte()
	}

	for c >= '0' && c <= '9' {
		val = val*10 + int(c-'0')

		nc, err := fs.r.ReadByte()
		if err != nil {
			break
		}
		c = nc
	}

	fs.r.UnreadByte()
	return sign * val
}

type Pair struct {
	node int
	dist int
}

var (
	N, K        int
	W           []int
	g           [][]int
	dist        []int
	visitedComp []bool
	queue       []int
)

func bfs(start, limit int, markComponent bool) Pair {

	head, tail := 0, 0
	queue[tail] = start
	tail++

	for i := 1; i <= N; i++ {
		dist[i] = -1
	}

	dist[start] = 0

	if markComponent {
		visitedComp[start] = true
	}

	farNode := start
	farDist := 0

	for head < tail {
		u := queue[head]
		head++

		if dist[u] > farDist {
			farDist = dist[u]
			farNode = u
		}

		for _, v := range g[u] {

			if W[v] > limit || dist[v] != -1 {
				continue
			}

			dist[v] = dist[u] + 1
			queue[tail] = v
			tail++

			if markComponent {
				visitedComp[v] = true
			}
		}
	}

	return Pair{farNode, farDist}
}

func check(limit int) bool {

	for i := 1; i <= N; i++ {
		visitedComp[i] = false
	}

	for i := 1; i <= N; i++ {

		if W[i] > limit || visitedComp[i] {
			continue
		}

		p1 := bfs(i, limit, true)
		p2 := bfs(p1.node, limit, false)

		if p2.dist >= K {
			return true
		}
	}

	return false
}

func main() {

	fs := NewFastScanner()

	N = fs.NextInt()
	K = fs.NextInt()

	W = make([]int, N+1)

	maxW := 0

	for i := 1; i <= N; i++ {
		W[i] = fs.NextInt()
		if W[i] > maxW {
			maxW = W[i]
		}
	}

	g = make([][]int, N+1)

	for i := 0; i < N-1; i++ {
		u := fs.NextInt()
		v := fs.NextInt()

		g[u] = append(g[u], v)
		g[v] = append(g[v], u)
	}

	dist = make([]int, N+1)
	visitedComp = make([]bool, N+1)
	queue = make([]int, N+5)

	low, high := 1, maxW
	ans := -1

	for low <= high {

		mid := low + (high-low)/2

		if check(mid) {
			ans = mid
			high = mid - 1
		} else {
			low = mid + 1
		}
	}

	fmt.Println(ans)
}

File: graphs/binary search concept/Tree Path.swift
import Foundation

final class FastScanner {
    private var data:[UInt8] = Array(FileHandle.standardInput.readDataToEndOfFile()) + [0]
    private var idx: Int = 0

    func nextInt() -> Int {
        while data[idx] == 10 || data[idx] == 13 || data[idx] == 32 || data[idx] == 9 {
            idx += 1
        }

        var sign = 1
        if data[idx] == 45 {
            sign = -1
            idx += 1
        }

        var num = 0
        while data[idx] >= 48 && data[idx] <= 57 {
            num = num * 10 + Int(data[idx] - 48)
            idx += 1
        }

        return num * sign
    }
}

struct Pair {
    var node: Int
    var dist: Int
}

let fs = FastScanner()

let N = fs.nextInt()
let K = fs.nextInt()

var W = Array(repeating: 0, count: N + 1)
var maxW = 0

for i in 1...N {
    W[i] = fs.nextInt()
    maxW = max(maxW, W[i])
}

var g = Array(repeating: [Int](), count: N + 1)

if N > 1 {
    for _ in 0..<(N - 1) {
        let u = fs.nextInt()
        let v = fs.nextInt()

        g[u].append(v)
        g[v].append(u)
    }
}

var dist = Array(repeating: -1, count: N + 1)
var visitedComp = Array(repeating: false, count: N + 1)
var queue = Array(repeating: 0, count: N + 5)

func bfs(_ start: Int, _ limit: Int, _ markComponent: Bool) -> Pair {

    for i in 1...N {
        dist[i] = -1
    }

    var head = 0
    var tail = 0

    queue[tail] = start
    tail += 1

    dist[start] = 0

    if markComponent {
        visitedComp[start] = true
    }

    var farNode = start
    var farDist = 0

    while head < tail {

        let u = queue[head]
        head += 1

        if dist[u] > farDist {
            farDist = dist[u]
            farNode = u
        }

        for v in g[u] {

            if W[v] > limit || dist[v] != -1 {
                continue
            }

            dist[v] = dist[u] + 1

            queue[tail] = v
            tail += 1

            if markComponent {
                visitedComp[v] = true
            }
        }
    }

    return Pair(node: farNode, dist: farDist)
}

func check(_ limit: Int) -> Bool {

    for i in 1...N {
        visitedComp[i] = false
    }

    for i in 1...N {

        if W[i] > limit || visitedComp[i] {
            continue
        }

        let p1 = bfs(i, limit, true)
        let p2 = bfs(p1.node, limit, false)

        if p2.dist >= K {
            return true
        }
    }

    return false
}

var low = 1
var high = maxW
var ans = -1

while low <= high {

    let mid = low + (high - low) / 2

    if check(mid) {
        ans = mid
        high = mid - 1
    } else {
        low = mid + 1
    }
}

print(ans)
